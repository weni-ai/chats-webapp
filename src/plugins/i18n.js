import Vue from 'vue';
import VueI18n from 'vue-i18n';

// eslint-disable-next-line camelcase
import pt_br from '@/locales/pt_br.json';
import en from '@/locales/en.json';
import es from '@/locales/es.json';

import moment from 'moment';

import('moment/locale/pt-br');
import('moment/locale/es');

moment.locale('pt-br');

Vue.use(VueI18n);

const messages = {
  'pt-br': pt_br,
  en,
  es,
};

const i18n = new VueI18n({
  locale: 'pt-br',
  fallbackLocale: 'pt-br',
  messages,
  dateTimeFormats: {
    'pt-br': {
      short: {
        year: 'numeric',
        month: '2-digit',
        day: '2-digit',
      },
      long: {
        month: 'long',
        day: '2-digit',
      },
    },
    en: {
      short: {
        year: 'numeric',
        month: '2-digit',
        day: '2-digit',
      },
      long: {
        month: 'long',
        day: '2-digit',
      },
    },
    es: {
      short: {
        year: 'numeric',
        month: '2-digit',
        day: '2-digit',
      },
      long: {
        month: 'long',
        day: '2-digit',
      },
    },
  },
});

export default i18n;
