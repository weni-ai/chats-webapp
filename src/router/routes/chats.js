import isMobile from 'is-mobile';

import DesktopHome from '@/views/chats/Home/index.vue';
import MobileHome from '@/views/chats/Mobile/MobileHome.vue';

const Home = isMobile() ? MobileHome : DesktopHome;

const routes = [
  {
    path: '/',
    name: 'root',
    beforeEnter: (from, to, next) => {
      if (to.path === '/') {
        if (isMobile()) {
          next({ name: 'orgs', replace: true });
        } else {
          if (from.query.next) {
            const isGeneralSettings = from.query.next === '/settings/chats';
            next({ path: isGeneralSettings ? '/settings' : from.query.next });
          } else next({ name: 'home', replace: true });
        }
      } else next(to.path);
    },
  },
  {
    path: '/:pathMatch(.*)*',
    name: '404',
    redirect: () => {
      return { path: `/` };
    },
  },
  {
    path: '/rooms',
    name: 'home',
    component: Home,
    props: (route) => ({
      newQuickMessage: route.query.newQuickMessage,
    }),
  },
  {
    path: '/chats/:roomId',
    name: 'room',
    component: Home,
    props: (route) => ({
      roomId: route.params.roomId,
    }),
  },
  {
    path: '/discussions/:discussionId',
    name: 'discussion',
    component: Home,
    props: (route) => ({
      discussionId: route.params.discussionId,
    }),
  },
  {
    path: '/closed-chats',
    name: 'closed-rooms',
    component: () => import('@/views/chats/ClosedChats/index.vue'),
    props: (route) => ({
      contactUrn: route.query.contactUrn,
      sector: route.query.sector,
      tags: route.query.tags,
      startDate: route.query.startDate,
      endDate: route.query.endDate,
      from: route.query.from,
    }),
  },
  {
    path: '/closed-chats/:roomId',
    name: 'closed-rooms.selected',
    component: () => import('@/views/chats/ClosedChats/index.vue'),
    props: (route) => ({
      roomId: route.params.roomId,
    }),
  },
];

export default routes;
