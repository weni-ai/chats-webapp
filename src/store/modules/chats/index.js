import quickMessages from './quickMessages';
import quickMessagesShared from './quickMessagesShared';

const tags = [
  { text: 'Dúvidas', value: 'doubts' },
  { text: 'Financeiro', value: 'finance' },
  { text: 'Ajuda', value: 'help' },
];
let chats = [
  {
    name: 'Fila',
    chats: [
      {
        id: '1',
        username: 'João Cláudio',
        lastMessage: 'Acho que podemos conversar ama...',
        unreadMessages: 8,
        waitingTime: 5,
        messages: [],
      },
      {
        id: '2',
        username: 'Marcela Barros',
        lastMessage: 'Acho que podemos conversar ama...',
        waitingTime: 3,
        unreadMessages: 5,
        messages: [],
      },
      {
        id: '3',
        username: 'Roberto Barcelos',
        lastMessage: 'Acho que podemos conversar ama...',
        waitingTime: 2,
        unreadMessages: 10,
        messages: [],
      },
    ],
  },
  {
    name: 'Chats Abertos',
    chats: [
      {
        id: '4',
        username: 'Pedro Pontes',
        lastMessage: 'Acho que podemos conversar ama...',
        unreadMessages: 8,
        messages: [
          {
            id: '1',
            username: 'Pedro Pontes',
            time: '10h15',
            content: [
              { text: 'Oi' },
              {
                text: 'Gostaria de saber se consigo pagar meu boleto depois do vencimento',
              },
              { text: 'O boleto venceu ontem e eu não queria pagar juros' },
              { text: 'Como faço?' },
            ],
          },
          {
            id: '2',
            username: 'Agente',
            time: '10h20',
            content: [
              { text: 'Olá, Pedro' },
              {
                text: 'Preciso confirmar com o financeiro, você me dá um minuto?',
              },
            ],
          },
          {
            id: '3',
            username: 'Pedro Pontes',
            time: '10h15',
            content: [
              { text: 'Oi' },
              {
                text: 'Gostaria de saber se consigo pagar meu boleto depois do vencimento',
              },
              { text: 'O boleto venceu ontem e eu não queria pagar juros' },
              { text: 'Como faço?' },
            ],
          },
          {
            id: '4',
            username: 'Agente',
            time: '10h20',
            content: [
              { text: 'Olá, Pedro' },
              {
                text: 'Preciso confirmar com o financeiro, você me dá um minuto?',
              },
            ],
          },
          {
            id: '5',
            username: 'Pedro Pontes',
            time: '10h15',
            content: [
              { text: 'Oi' },
              {
                text: 'Gostaria de saber se consigo pagar meu boleto depois do vencimento',
              },
              { text: 'O boleto venceu ontem e eu não queria pagar juros' },
              { text: 'Como faço?' },
            ],
          },
          {
            id: '6',
            username: 'Agente',
            time: '10h20',
            content: [
              { text: 'Olá, Pedro' },
              {
                text: 'Preciso confirmar com o financeiro, você me dá um minuto?',
              },
            ],
          },
        ],
      },
      {
        id: '5',
        username: 'Ilanna Lins',
        lastMessage: 'Acho que podemos conversar ama...',
        unreadMessages: 5,
        messages: [
          {
            id: '1',
            username: 'Ilanna Lins',
            time: '10h15',
            content: [
              { text: 'Oi' },
              {
                text: 'Gostaria de saber como faço para alterar as configurações do meu perfil',
              },
            ],
          },
          {
            id: '2',
            username: 'Agente',
            time: '10h20',
            content: [
              { text: 'Olá, Ilanna' },
              {
                text: 'Vou te mostrar como fazer, é só acessar seu perfil e seguir os passos contidos na imagem',
              },
              {
                // Get a random image with 1920x1080 pixels (16:9 proportion)
                src: 'https://picsum.photos/1920/1080',
                type: 'image',
                isMedia: true,
                fileExtension: 'png',
                filename: 'Captura de tela',
              },
            ],
          },
          {
            id: '3',
            username: 'Ilanna Lins',
            time: '10h15',
            content: [
              { text: 'Muito obrigada!' },
              {
                text: 'Me ajudou muito :)',
              },
            ],
          },
        ],
      },
      {
        id: '6',
        username: 'Rafael Felipe',
        lastMessage: 'Acho que podemos conversar ama...',
        unreadMessages: 10,
        messages: [
          {
            id: '1',
            username: 'Rafael Felipe',
            time: '10h15',
            content: [
              { text: 'Oi' },
              {
                text: 'Gostaria de saber como faço para alterar as configurações do meu perfil',
              },
            ],
          },
          {
            id: '2',
            username: 'Agente',
            time: '10h20',
            content: [
              { text: 'Olá, Rafael' },
              {
                text: 'Vou te mostrar como fazer, é só acessar seu perfil e seguir os passos contidos que o vídeo mostra:',
              },
              {
                // example video
                src: 'https://test-videos.co.uk/vids/bigbuckbunny/mp4/h264/1080/Big_Buck_Bunny_1080_10s_1MB.mp4',
                type: 'video',
                isMedia: true,
                fileExtension: 'mp4',
                filename: 'Tutorial de cadastro',
              },
            ],
          },
          {
            id: '3',
            username: 'Rafael Felipe',
            time: '10h15',
            content: [
              { text: 'Muito obrigada!' },
              {
                text: 'Me ajudou muito :)',
              },
            ],
          },
        ],
      },
      {
        id: '8',
        username: 'Milton Neves',
        lastMessage: 'Acho que podemos conversar ama...',
        unreadMessages: 3,
        messages: [
          {
            id: '1',
            username: 'Milton Neves',
            time: '10h15',
            content: [
              { text: 'Olá' },
              {
                text: 'Gostaria de uma segunda via da minha fatura com prazo atualizado',
              },
            ],
          },
          {
            id: '2',
            username: 'Agente',
            time: '10h20',
            content: [
              { text: 'Olá, Milton' },
              {
                text: 'Segue o boleto atualizado!',
              },
              {
                // image with A4 proportion to simulate a PDF document
                src: 'https://picsum.photos/2480/3508',
                type: 'document',
                isMedia: true,
                fileExtension: 'pdf',
                filename: 'Boleto atualizado',
              },
            ],
          },
          {
            id: '3',
            username: 'Milton Neves',
            time: '10h22',
            content: [
              { text: 'Muito obrigado!' },
              {
                text: 'Me ajudou muito :)',
              },
              {
                text: 'Tenha um bom dia!',
              },
            ],
          },
        ],
      },
    ],
  },
];

const closedChats = [
  {
    id: '1',
    username: 'Milton Neves',
    agent: 'Debora',
    date: '25/03/2022',
    closed: true,
    tags: [tags[1], tags[2]],
    rooms: [
      {
        id: '1',
        agent: 'Bot',
        transferredTo: 'fila Financeiro',
        messages: [
          {
            id: '1',
            username: 'Milton Neves',
            time: '10h15',
            content: [{ text: 'Olá estou precisando de ajuda' }],
          },
          {
            id: '2',
            username: 'Bot',
            time: '10h18',
            content: [{ text: 'Olá Milton, como posso te ajudar?' }],
          },
          {
            id: '3',
            username: 'Milton Neves',
            time: '10h20',
            content: [{ text: 'Preciso da segunda via de um boleto' }],
          },
          {
            id: '4',
            username: 'Bot',
            time: '10h20',
            content: [{ text: 'Vou precisar que você confirme seu CPF' }],
          },
          {
            id: '6',
            username: 'Milton Neves',
            time: '10h21',
            content: [{ text: '234.343.348-95' }],
          },
          {
            id: '7',
            username: 'Bot',
            time: '10h21',
            content: [
              { text: 'Milton, vou te passar para o atendimento humano, peço que aguarde' },
            ],
          },
        ],
      },
      {
        id: '2',
        agent: 'Fabricio Correia',
        transferredTo: 'fila Suporte',
        messages: [
          {
            id: '1',
            username: 'Fabricio Correia',
            time: '10h18',
            content: [{ text: 'Olá Milton, aguarda um pouquinho que te encaminho o boleto' }],
          },
          {
            id: '3',
            username: 'Milton Neves',
            time: '10h20',
            content: [{ text: 'Claro! fico no aguardo' }],
          },
          {
            id: '4',
            username: 'Fabricio Correia',
            time: '10h20',
            content: [{ text: 'Posso te ajudar em algo mais Milton?' }],
          },
          {
            id: '6',
            username: 'Milton Neves',
            time: '10h21',
            content: [{ text: 'Estou com uma dificuldade no meu aplicativo' }],
          },
          {
            id: '7',
            username: 'Fabricio Correia',
            time: '10h20',
            content: [
              {
                text: 'Olá Ilanna, nesse caso vou te passar para o setor de suporte, só aguarda um pouquinho',
              },
            ],
          },
          {
            id: '8',
            username: 'Milton Neves',
            time: '10h21',
            content: [{ text: 'ok aguardo' }],
          },
        ],
      },
      {
        id: '3',
        agent: 'Amanda Santos',
        closedBy: 'agent',
        tags: [tags[1], tags[2]],
        messages: [
          {
            id: '1',
            username: 'Amanda Santos',
            time: '10h18',
            content: [{ text: 'Olá Milton, aguarda um pouquinho que te encaminho o boleto' }],
          },
          {
            id: '3',
            username: 'Milton Neves',
            time: '10h20',
            content: [{ text: 'Claro! fico no aguardo' }],
          },
          {
            id: '4',
            username: 'Amanda Santos',
            time: '10h20',
            content: [{ text: 'Posso te ajudar em algo mais Ilanna?' }],
          },
          {
            id: '6',
            username: 'Milton Neves',
            time: '10h21',
            content: [{ text: 'Estou com uma dificuldade no meu aplicativo' }],
          },
          {
            id: '7',
            username: 'Amanda Santos',
            time: '10h20',
            content: [
              {
                text: 'Olá Ilanna, nesse caso vou te passar para o setor de suporte, só aguarda um pouquinho',
              },
            ],
          },
          {
            id: '8',
            username: 'Milton Neves',
            time: '10h21',
            content: [{ text: 'ok aguardo' }],
          },
        ],
      },
    ],
  },
  {
    id: '2',
    username: 'João Claudio',
    date: '05/02/2022',
    tags: [tags[1]],
    agent: 'Felipe',
    closed: true,
    rooms: [
      {
        agent: 'Felipe',
        tags: [tags[1]],
        closedBy: 'agent',
        messages: [
          {
            id: '1',
            username: 'João Claudio',
            time: '10h15',
            content: [
              { text: 'Oi' },
              {
                text: 'Gostaria de saber se consigo pagar meu boleto depois do vencimento',
              },
              { text: 'O boleto venceu ontem e eu não queria pagar juros' },
              { text: 'Como faço?' },
            ],
          },
          {
            id: '2',
            username: 'Agente',
            time: '10h20',
            content: [
              { text: 'Olá, João' },
              {
                text: 'Preciso confirmar com o financeiro, você me dá um minuto?',
              },
            ],
          },
        ],
      },
    ],
  },
  {
    id: '3',
    username: 'Dulce Figueira',
    agent: 'Ana',
    date: '02/02/2022',
    closed: true,
    tags: [tags[0]],
    rooms: [
      {
        agent: 'Ana',
        closedBy: 'agent',
        tags: [tags[0]],
        messages: [
          {
            id: '1',
            username: 'Dulce Figueira',
            time: '10h15',
            content: [
              { text: 'Oi' },
              {
                text: 'Gostaria de saber se consigo pagar meu boleto depois do vencimento',
              },
              { text: 'O boleto venceu ontem e eu não queria pagar juros' },
              { text: 'Como faço?' },
            ],
          },
          {
            id: '2',
            username: 'Agente',
            time: '10h20',
            content: [
              { text: 'Olá, Dulce' },
              {
                text: 'Preciso confirmar com o financeiro, você me dá um minuto?',
              },
            ],
          },
        ],
      },
    ],
  },
];

const channels = ['WhatsApp', 'Telegram', 'Instagram'];

chats = chats.map((group) => ({
  ...group,
  chats: group.chats.map((chat) => {
    const oneMinuteInMs = 60 * 1000;
    const oneHourInMs = 60 * oneMinuteInMs;
    const oneDayInMs = 24 * oneHourInMs;

    const today = new Date();
    const yesterday = new Date(today.getTime() - oneDayInMs);

    const getRandomDate = () => {
      const fewMinutesAgo = Math.floor(Math.random() * (59 - 1) + 1);
      const fewHoursAgo = Math.floor(Math.random() * (23 - 1) + 1);

      const date =
        today.getTime() -
        (Math.random() > 0.5 ? fewHoursAgo * oneHourInMs : fewMinutesAgo * oneMinuteInMs);
      return new Date(date);
    };

    const wasLastViewYesterday = Math.random() > 0.5;

    const randomDates = [getRandomDate(), getRandomDate()];
    randomDates.sort((a, b) => b - a);

    const lastView = wasLastViewYesterday ? yesterday : randomDates[0];
    const lastContactDate = wasLastViewYesterday ? yesterday : randomDates[1];

    const channel = channels[Math.round(Math.random() * (channels.length - 1))];

    return {
      ...chat,
      lastView,
      lastContactDate,
      channel,
    };
  }),
}));

const module = {
  namespaced: true,
  modules: {
    quickMessages,
    quickMessagesShared,
  },
  state: {
    activeChat: null,
    chats,
    tags,
    closedChats,
  },

  mutations: {
    addClosedChat(state, chat) {
      state.closedChats.unshift(chat);
    },
    removeChat(state, chatToRemove) {
      state.chats = state.chats.map((group) => ({
        ...group,
        chats: group.chats.filter((c) => c.id !== chatToRemove.id),
      }));
    },
    reorderChats(state, activeChat) {
      const activeChatGroup = state.chats.find(
        (group) => !!group.chats.find((chat) => chat.id === activeChat.id),
      );

      activeChatGroup.chats = activeChatGroup.chats.filter((chat) => chat.id !== activeChat.id);
      activeChatGroup.chats.unshift(activeChat);
    },
    setActiveChat(state, chat) {
      state.activeChat = chat ? { ...chat, unreadMessages: 0 } : null;

      if (!chat) return;

      state.chats = state.chats.map((group) => ({
        ...group,
        chats: group.chats.map((c) => (c.id === chat.id ? { ...c, unreadMessages: 0 } : c)),
      }));
    },
    sendMessage(state, message) {
      if (!state.activeChat) return;

      const { messages } = state.activeChat;

      if (messages.at(-1)?.username === 'Agente') {
        messages.at(-1).content.push(message);
      } else {
        messages.push({
          id: Math.ceil(Math.random() * 100 + 1),
          username: 'Agente',
          time: `${new Date().getHours().toString().padStart(2, '0')}h${new Date()
            .getMinutes()
            .toString()
            .padStart(2, '0')}`,
          content: [message],
        });
      }

      state.activeChat.messages = messages;
    },
  },

  getters: {
    getChatById(state) {
      const chats = state.chats.map((chatGroup) => chatGroup.chats).flat();

      return (id) => chats.find((chat) => chat.id === id.toString());
    },
  },

  actions: {
    closeChat({ commit }, chat) {
      commit('removeChat', chat);

      const formattedChat = {
        ...chat,
        rooms: [
          {
            agent: chat.agent,
            messages: chat.messages,
            tags: chat.tags,
            closedBy: 'agent',
          },
        ],
      };

      commit('addClosedChat', formattedChat);
    },
    sendMessage({ commit, state }, content) {
      if (!state.activeChat) return;

      const message =
        typeof content === 'string' ? { text: content, sent: Math.random() < 0.1 } : content;

      commit('sendMessage', message);
      commit('reorderChats', state.activeChat);
    },
  },
};

export default module;
