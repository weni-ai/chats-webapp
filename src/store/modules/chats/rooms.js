import { defineStore } from 'pinia';

import { useDashboard } from '../dashboard';
import { useProfile } from '../profile';

import Room from '@/services/api/resources/chats/room';

export const useRooms = defineStore('rooms', {
  state: () => ({
    rooms: [],
    activeRoom: null,
    newMessagesByRoom: {},
    hasNextRooms: true,
    canUseCopilot: false,
    copilotSuggestion: '',

    selectedRoomsToTransfer: [],
    contactToTransfer: '',
  }),

  actions: {
    updateMessagesByRoom({ room, message, reset = false }) {
      const roomMessages = this.newMessagesByRoom[room]?.messages || [];

      this.newMessagesByRoom = {
        ...this.newMessagesByRoom,
        [room]: {
          messages: reset ? [] : [...roomMessages, message],
        },
      };
    },

    setActiveRoom(room) {
      this.activeRoom = room;
    },

    addRoom(room) {
      if (room.uuid) {
        this.rooms.unshift({ ...room });
      }
    },

    bringRoomFront(room) {
      this.rooms.sort((x) => (x === room ? -1 : 0));
    },

    setCopilotSuggestion(suggestion) {
      this.copilotSuggestion = suggestion;
    },

    checkUserSeenRoom({ room, userEmail, viewedAgentEmail }) {
      const profileStore = useProfile();
      const isProjectAdmin = profileStore.me.project_permission_role === 1;

      if (isProjectAdmin && !room.user) return true;

      const userHasRoomQueue = !!profileStore.me.queues?.find(
        (permission) =>
          permission.queue === room.queue.uuid && permission.role === 1,
      );

      if (!room.user && userHasRoomQueue) return true;

      if (viewedAgentEmail) {
        return room.user?.email === viewedAgentEmail;
      }

      return room.user?.email === userEmail;
    },

    async getAll({ offset, concat, limit, contact, order, viewedAgent }) {
      const response = await Room.getAll(
        offset,
        limit,
        contact,
        order,
        viewedAgent,
      );
      let gettedRooms = response.results || [];
      const listRoomHasNext = response.next;
      if (concat) {
        gettedRooms = this.rooms.concat(response.results);
      }
      this.hasNextRooms = listRoomHasNext;
      this.rooms = gettedRooms;

      return gettedRooms;
    },

    async updateRoomContact({ uuid }) {
      const newRoom = await Room.getByUuid({ uuid });

      this.activeRoom = newRoom;
    },

    async getCanUseCopilot() {
      if (this.activeRoom) {
        const response = await Room.getCanUseCopilot({
          uuid: this.activeRoom.uuid,
        });
        this.canUseCopilot = response.can_use_chat_completion;
      }
    },

    async getCopilotSuggestion() {
      this.setCopilotSuggestion('');
      const response = await Room.getCopilotSuggestion({
        uuid: this.activeRoom.uuid,
      });
      const suggestion = response?.choices?.[0]?.message?.content;
      if (suggestion) {
        this.setCopilotSuggestion(suggestion || '');
      } else if (response.status !== 200) {
        return response.status;
      }
      return undefined;
    },

    updateRoom({ room, userEmail, routerReplace, viewedAgentEmail }) {
      const dashboardStore = useDashboard();
      const filteredRooms = this.rooms
        .map((mappedRoom) =>
          mappedRoom.uuid === room.uuid ? { ...room } : mappedRoom,
        )
        .filter((filteredRoom) => {
          return this.checkUserSeenRoom({
            room: filteredRoom,
            viewedAgentEmail,
            userEmail,
          });
        });

      this.rooms = filteredRooms;

      const isTransferedToOtherUser =
        room.user && room.user.email !== userEmail;

      const isTransferedByMe = room.transferred_by === userEmail;

      const isTransferedByViewedAgent =
        room.transferred_by === viewedAgentEmail;

      const isTransferedFromAQueue =
        room.transfer_history?.from?.type === 'queue' ||
        !room.transfer_history?.from;

      const isActiveRoom =
        this.activeRoom && room.uuid === this.activeRoom.uuid;

      if (!isTransferedByMe && isTransferedToOtherUser) {
        if (!isTransferedFromAQueue && !room.is_waiting && !viewedAgentEmail) {
          dashboardStore.setShowModalAssumedChat(true);
          dashboardStore.setAssumedChatContactName(room.contact.name);
        }

        if (isActiveRoom && !viewedAgentEmail) {
          this.setActiveRoom(null);
          routerReplace();
          return;
        }
      }

      if (!room.is_waiting && isActiveRoom) {
        if (isTransferedByViewedAgent) {
          this.setActiveRoom(null);
          return;
        }
        this.setActiveRoom({ ...room });
      }
    },

    removeRoom(roomUuid) {
      const filteredRooms = this.rooms.filter((r) => r.uuid !== roomUuid);
      this.rooms = filteredRooms;

      if (this.activeRoom && this.activeRoom?.uuid === roomUuid) {
        this.setActiveRoom(null);
      }
    },

    addNewMessagesByRoom({ room, message }) {
      this.updateMessagesByRoom({ room, message });
    },

    resetNewMessagesByRoom({ room }) {
      this.updateMessagesByRoom({ room, reset: true });
    },

    setSelectedRoomsToTransfer(rooms) {
      this.selectedRoomsToTransfer = rooms;
    },

    setContactToTransfer(contact) {
      this.contactToTransfer = contact;
    },
  },

  getters: {
    agentRooms(store) {
      return store.rooms.filter(
        (room) => !!room.user && room.is_waiting === false,
      );
    },
    waitingQueue(store) {
      return store.rooms.filter((room) => !room.user && !room.is_waiting);
    },
    waitingContactAnswer(store) {
      return store.rooms.filter((room) => room.is_waiting === true);
    },
    getRoomById: (store) => (uuid) => {
      return store.rooms.find((room) => room.uuid === uuid);
    },
  },
});
