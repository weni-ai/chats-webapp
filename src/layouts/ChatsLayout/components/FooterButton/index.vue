<template>
  <section
    class="chats-layout-footer-button"
    data-testid="footer-button-container"
  >
    <section
      v-if="selectedRoomsToTransfer.length >= 1"
      class="chats-layout-footer-button__bulk-transfer"
      data-testid="bulk-transfer-section"
    >
      <UnnnicButton
        class="chats-layout-footer-button__button"
        :text="$tc('transfer_contact', selectedRoomsToTransfer.length)"
        iconLeft="read_more"
        type="primary"
        size="small"
        data-testid="transfer-button"
        @click="handleModalBulkTransfer"
      />
      <ModalBulkTransfer
        v-if="isModalBulkTransferOpened"
        data-testid="bulk-transfer-modal"
        @close="handleModalBulkTransfer"
      />
    </section>
  </section>
</template>

<script>
import { mapState } from 'pinia';

import { useRooms } from '@/store/modules/chats/rooms';

import ModalBulkTransfer from '@/components/chats/chat/ModalBulkTransfer.vue';
export default {
  name: 'ChatsLayoutFooterButton',

  components: {
    ModalBulkTransfer,
  },

  data() {
    return {
      isModalBulkTransferOpened: false,
    };
  },

  computed: {
    ...mapState(useRooms, ['selectedRoomsToTransfer']),
  },
  methods: {
    handleModalBulkTransfer() {
      this.isModalBulkTransferOpened = !this.isModalBulkTransferOpened;
    },
  },
};
</script>

<style lang="scss" scoped>
.chats-layout-footer-button {
  padding-left: $unnnic-spacing-xs;
  &__button {
    width: 100%;
  }
}
</style>
