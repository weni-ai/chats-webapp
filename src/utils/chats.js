import { useDiscussions } from '@/store/modules/chats/discussions';
import { useDiscussionMessages } from '@/store/modules/chats/discussionMessages';
import { useRoomMessages } from '@/store/modules/chats/roomMessages';
import { useRooms } from '@/store/modules/chats/rooms';

export async function resetChats() {
  await useDiscussionMessages().resetDiscussionMessages();
  await useRoomMessages().resetRoomMessages();
  await useDiscussions().setActiveDiscussion(null);
  await useRooms().setActiveRoom(null);
}

export function removeDuplicatedMessagesByUuid(arrayMessages) {
  const itemsChecked = new Set();
  return arrayMessages.filter((item) => {
    if (itemsChecked.has(item.uuid)) {
      return false;
    }
    itemsChecked.add(item.uuid);
    return true;
  });
}

export function formatContactName(room) {
  let contactName = '';
  if (room?.service_chat) {
    contactName += `${room.service_chat} | `;
  }
  if (room?.protocol) {
    contactName += `${room.protocol} | `;
  }
  return (contactName + room?.contact.name).trim();
}
