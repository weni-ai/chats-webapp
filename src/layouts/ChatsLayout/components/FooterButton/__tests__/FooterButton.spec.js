import { afterAll, beforeAll, describe, expect, it } from 'vitest';
import { mount, config } from '@vue/test-utils';
import { createPinia, setActivePinia } from 'pinia';
import FooterButton from '@/layouts/ChatsLayout/components/FooterButton/index.vue';
import { useRooms } from '@/store/modules/chats/rooms';
import i18n from '@/plugins/i18n';

beforeAll(() => {
  config.global.plugins = config.global.plugins.filter(
    (plugin) => plugin !== i18n,
  );
});

afterAll(() => {
  if (!config.global.plugins.includes(i18n)) {
    config.global.plugins.push(i18n);
  }
});

describe('FooterButton', () => {
  const createWrapper = (selectedRooms = []) => {
    const pinia = createPinia();
    setActivePinia(pinia);
    const roomsStore = useRooms();
    roomsStore.selectedRoomsToTransfer = selectedRooms;

    return mount(FooterButton, {
      global: {
        plugins: [pinia],
        mocks: {
          $tc: (key, count) => `${key}_${count}`,
        },
        stubs: {
          UnnnicButton: true,
          ModalTransferRooms: true,
        },
      },
    });
  };

  describe('Initial State and Store Integration', () => {
    it('should initialize correctly and integrate with store', () => {
      const wrapper = createWrapper(['room1', 'room2']);

      expect(wrapper.vm.isModalTransferRoomsOpened).toBe(false);
      expect(wrapper.vm.selectedRoomsToTransfer).toEqual(['room1', 'room2']);
      expect(
        wrapper.find('[data-testid="footer-button-container"]').exists(),
      ).toBe(true);
    });
  });

  describe('Conditional Rendering', () => {
    it('should show bulk transfer section when rooms are selected', () => {
      const wrapper = createWrapper(['room1', 'room2', 'room3']);

      expect(
        wrapper.find('[data-testid="bulk-transfer-section"]').exists(),
      ).toBe(true);
      expect(wrapper.find('[data-testid="transfer-button"]').exists()).toBe(
        true,
      );
      expect(wrapper.findComponent({ name: 'UnnnicButton' }).exists()).toBe(
        true,
      );
    });

    it('should hide bulk transfer section when no rooms selected', () => {
      const wrapper = createWrapper([]);

      expect(
        wrapper.find('[data-testid="bulk-transfer-section"]').exists(),
      ).toBe(false);
      expect(wrapper.find('[data-testid="transfer-button"]').exists()).toBe(
        false,
      );
    });

    it('should conditionally show modal based on state', async () => {
      const wrapper = createWrapper(['room1']);

      expect(wrapper.find('[data-testid="bulk-transfer-modal"]').exists()).toBe(
        false,
      );

      await wrapper.setData({ isModalBulkTransferOpened: true });
      expect(wrapper.find('[data-testid="bulk-transfer-modal"]').exists()).toBe(
        true,
      );
    });
  });

  describe('Functionality', () => {
    it('should handle modal toggle method', () => {
      const wrapper = createWrapper(['room1']);

      expect(wrapper.vm.isModalTransferRoomsOpened).toBe(false);

      wrapper.vm.handleModalTransferRooms();
      expect(wrapper.vm.isModalTransferRoomsOpened).toBe(true);

      wrapper.vm.handleModalTransferRooms();
      expect(wrapper.vm.isModalTransferRoomsOpened).toBe(false);
    });

    it('should handle button click event through component interaction', async () => {
      const wrapper = createWrapper(['room1']);
      const button = wrapper.findComponent({ name: 'UnnnicButton' });

      expect(wrapper.vm.isModalTransferRoomsOpened).toBe(false);

      await button.vm.$emit('click');
      expect(wrapper.vm.isModalTransferRoomsOpened).toBe(true);
    });

    it('should handle modal close event', async () => {
      const wrapper = createWrapper(['room1']);

      wrapper.vm.handleModalTransferRooms();
      expect(wrapper.vm.isModalTransferRoomsOpened).toBe(true);

      await wrapper.vm.$nextTick();

      wrapper.vm.handleModalTransferRooms();
      expect(wrapper.vm.isModalTransferRoomsOpened).toBe(false);
    });

    it('should handle different room counts correctly', () => {
      const wrapperSingle = createWrapper(['room1']);
      expect(wrapperSingle.vm.selectedRoomsToTransfer).toHaveLength(1);

      const wrapperMultiple = createWrapper([
        'room1',
        'room2',
        'room3',
        'room4',
      ]);
      expect(wrapperMultiple.vm.selectedRoomsToTransfer).toHaveLength(4);

      expect(wrapperSingle.vm.$tc('transfer_contact', 1)).toBe(
        'transfer_contact_1',
      );
    });
  });

  describe('Snapshot', () => {
    it('should match snapshot', () => {
      const wrapper = createWrapper(['room1', 'room2']);
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
