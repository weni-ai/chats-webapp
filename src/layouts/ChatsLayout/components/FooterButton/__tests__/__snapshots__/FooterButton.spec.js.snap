// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FooterButton > Snapshot > should match snapshot 1`] = `
<section
  class="chats-layout-footer-button"
  data-testid="footer-button-container"
  data-v-944d9ccc=""
>
  <section
    class="chats-layout-footer-button__bulk-transfer"
    data-testid="bulk-transfer-section"
    data-v-944d9ccc=""
  >
    <unnnic-button-stub
      class="chats-layout-footer-button__button"
      data-testid="transfer-button"
      data-v-944d9ccc=""
      disabled="false"
      float="false"
      iconcenter=""
      iconleft="read_more"
      iconright=""
      iconsfilled="false"
      loading="false"
      next="false"
      size="small"
      text="transfer_contact_2"
      type="primary"
    />
    <!--v-if-->
  </section>
</section>
`;
